guardian/typerighter
# Create live and draft rules for dictionary words on ingest

## What does this change?

This PR converts dictionary words to live and draft rules on ingest of the dictionary file.

It also provides an interface for listing dictionary rules in the rule manager, this has been designed to work well with a debounced 'search-as-you-type' search similar to the one in the existing rules table:

`/api/dictionary` - > List all rules, showing the first 1000
`/api/dictionary?page=1` -> List all rules, showing the second 1000 (pages are zero-indexed)
`/api/dictionary?word=bel` -> show words starting with 'bel', ordered by word similarity (based on shared trigrams between search term and result word)
`/api/dictionary?word=a&page=2` -> show the third batch of 1000 words starting with 'a'

![image](https://github.com/guardian/typerighter/assets/34686302/a6703ddb-6cad-4d46-a712-c08ec7aaf5c5)

Dictionary rules are not included in the standard `list` endpoint so that they don't clutter the frontend rules table.

The endpoints are paginated for performance reasons - serialising to JSON all the rows for e.g. the search 'a' would be noticeably slow.

In follow-up PRs:
- Dictionary rules will be included in the published rules artefact and ingested by the Checker service, where they will be used in a spellchecker Matcher Pool (they're not included in this PR because they're not 'active' live rules and the there's some additional work required to include them in the artefact)
- We will introduce a frontend interface for listing, unpublishing and editing tags of dictionary rules

## How to test

1. Make sure you've run the 'setup' script to ensure the dictionary XML file is available locally.
2. Run the rule manager locally according to the instructions in the readme
3. Send a request to https://manager.typerighter.local.dev-gutools.co.uk/api/refreshDictionary in order to add the dictionary rules to your local database
4. Make some requests to `/api/dictionary`, e.g. `/api/dictionary?word=bel`. Do they behave as expected?
